subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java

```diff
@@ -211,34 +212,54 @@ public void visitProducerTasks(Action<? super Task> visitor) {
          * and instead always executes against the given {@code value}.
          */
         static <T, A> SideEffect<T> fixed(A value, SideEffect<A> sideEffect) {
-            return new FixedSideEffect<>(value, sideEffect);
+            return EmptySideEffect.isEmpty(sideEffect) ? EmptySideEffect.value() : new FixedSideEffect<>(value, sideEffect);
+        }
+
+        static <T, A> SideEffect<T> fixedFrom(Value<A> value) {
+            if (value.isMissing()) {
+                return EmptySideEffect.value();
+            }
+
+            SideEffect<? super A> sideEffect = value.getSideEffect();
+            return EmptySideEffect.isEmpty(sideEffect) ? EmptySideEffect.value() : fixed(value.getWithoutSideEffect(), sideEffect);
         }
 
         /**
          * Creates a new side effect that executes given side effects sequentially.
          */
-        static <T> SideEffect<T> composite(Iterable<SideEffect<? super T>> sideEffects) {
-            // TODO: optimize for empty and single-value cases
-            return new CompositeSideEffect<>(sideEffects);
+        static <T> SideEffect<T> composite(Iterable<SideEffect<T>> sideEffects) {
+            return CompositeSideEffect.of(sideEffects);
         }
 
         /**
          * Creates a new side effect that executes given side effects sequentially.
          */
         @SafeVarargs
         static <T> SideEffect<T> composite(SideEffect<? super T>... sideEffects) {
-            ArrayList<SideEffect<? super T>> flatSideEffects = new ArrayList<>(sideEffects.length);
-            for (SideEffect<? super T> sideEffect : sideEffects) {
-                if (sideEffect instanceof CompositeSideEffect) {
-                    CompositeSideEffect<? super T> compositeSideEffect = Cast.uncheckedNonnullCast(sideEffect);
-                    flatSideEffects.addAll(compositeSideEffect.sideEffects);
-                } else {
-                    flatSideEffects.add(sideEffect);
-                }
-            }
+            @SuppressWarnings("varargs")
+            List<SideEffect<T>> sideEffectList = Cast.uncheckedNonnullCast(Arrays.asList(sideEffects));
+            return CompositeSideEffect.of(sideEffectList);
+        }
+
+        static <T> ProviderInternal<T> attach(ProviderInternal<T> provider, @Nullable SideEffect<? super T> sideEffect) {
+            return EmptySideEffect.isEmpty(sideEffect) ? provider : provider.withSideEffect(sideEffect);
+        }
```

Comment:
I wonder if we should just expect a nullable `SideEffect` in the `withSideEffect` method for `Provider`/`Value`/`ExecutionTimeValue`. Then we will not need any of these `attach` helpers.

If the `withSideEffect` was exposed to the users, I would most likely ignored this idea, but since it is on the `ProviderInternal` it may make sense.

Reply:
You mean allowing `withSideEffect(null)` only makes sense if we remove `EmptySideEffect` too?

Reply:
I thought we could remove the empty side effect if we allow `withSideEffect(null)`. Is that not the case? How much simpler will the code become if we only allow `withSideEffect(null)`?

Reply:
I implemented the support for the null side effect arguments in a separate PR, so it's easier to discuss.

Looking at it, I like that it simplifies a number of places and allows to call `.withSideEffect(nullableSideEffect)` everywhere in the implementation.

https://github.com/gradle/gradle/pull/21674